torch/csrc/dynamo/eval_frame.c

```diff
@@ -304,10 +305,101 @@ typedef struct cache_entry {
   struct cache_entry* next;
 } CacheEntry;
 
+static void cache_entry_dealloc(CacheEntry* e);
+
+#define DECLARE_CACHE_ENTRY_ATTR(name) \
+static PyObject* CacheEntry_##name(CacheEntry* self, PyObject* _noargs) { \
+  PyObject* res = (PyObject*)self->name; \
+  Py_INCREF(res); \
+  return res; \
+}
+
+DECLARE_CACHE_ENTRY_ATTR(check_fn)
+DECLARE_CACHE_ENTRY_ATTR(code)
+DECLARE_CACHE_ENTRY_ATTR(next)
+
+static struct PyGetSetDef CacheEntry_properties[] = {
+    {"check_fn", (getter)CacheEntry_check_fn, NULL, NULL, NULL},
+    {"code", (getter)CacheEntry_code, NULL, NULL, NULL},
+    {"next", (getter)CacheEntry_next, NULL, NULL, NULL},
+    {NULL}};
+
+
+static PyObject* cache_entry_new(PyTypeObject* type, PyObject* args, PyObject* kwargs) {
+  CacheEntry *self;
+  self = (CacheEntry*) type->tp_alloc(type, 0);
+  if (self != NULL) {
+    // The corresponding decrefs for Py_None are in cache_entry_init.
+    Py_INCREF(Py_None);
+    self->check_fn = Py_None;
+    Py_INCREF(Py_None);
+    self->code = (PyCodeObject*)Py_None;
+    Py_INCREF(Py_None);
+    self->next = (CacheEntry*)Py_None;
+  }
+  return (PyObject*)self;
+}
+
+
+static int cache_entry_init(CacheEntry* self, PyObject* args, PyObject* kwds) {
+  PyObject* check_fn = NULL;
+  PyCodeObject* code = NULL;
+  CacheEntry* next = NULL;
+
+  static char *kwlist[] = {"check_fn", "code", "next", NULL};
+
+  int ret = PyArg_ParseTupleAndKeywords(
+    args, kwds, "OOO", kwlist,
+    &check_fn, &code, &next);
+
+  if (!ret) return -1;
+
+  if (check_fn) {
+    PyObject* tmp = self->check_fn;
+    Py_INCREF(check_fn);
+    self->check_fn = check_fn;
+    Py_XDECREF(tmp);
+  }
+
+  if (code) {
+    PyCodeObject* tmp = self->code;
+    Py_INCREF(code);
+    self->code = code;
+    Py_XDECREF(tmp);
+  }
+
+  if (next) {
+    CacheEntry* tmp = self->next;
+    Py_INCREF(next);
+    self->next = next;
+    Py_XDECREF(tmp);
+  }
```

Comment:
@ezyang One last review of these lines and the corresponding lines in `cache_entry_init`.